tutorials/optimization/gpu_optimization.rst

```diff
@@ -0,0 +1,258 @@
+particular may struggle with scenes that run easily on desktop.
+
+Understanding and investigating GPU bottlenecks is slightly different to the
+situation on the CPU, because often you can only change performance indirectly,
+by changing the instructions you give to the GPU, and it may be more difficult
+to take measurements. Often the only way of measuring performance is by
+examining changes in frame rate.
+
+Drawcalls, state changes, and APIs
+==================================
+
+.. note:: The following section is not relevant to end-users, but is useful to
+          provide background information that is relevant in later sections.
+
+Godot sends instructions to the GPU via a graphics API (OpenGL, GLES2, GLES3,
+Vulkan). The communication and driver activity involved can be quite costly,
+especially in OpenGL. If we can provide these instructions in a way that is
+preferred by the driver and GPU, we can greatly increase performance.
+
+Nearly every API command in OpenGL requires a certain amount of validation, to
+make sure the GPU is in the correct state. Even seemingly simple commands can
+lead to a flurry of behind the scenes housekeeping. Therefore the name of the
+game is reduce these instructions to a bare minimum, and group together similar
+objects as much as possible so they can be rendered together, or with the
+minimum number of state changes.
+
+2D batching
+~~~~~~~~~~~
+
+In 2d, the costs of treating each item individually can be prohibitively high -
+there can easily be thousands on screen. This is why 2d batching is used -
+multiple similar items are grouped together and rendered in a batch, via a
+single drawcall, rather than making a separate drawcall for each item. In
+addition this means that state changes, material and texture changes can be kept
+to a minimum.
+
+For more information on 2D batching see :ref:`doc_batching`.
+
+3D batching
+~~~~~~~~~~~
+
+In 3d, we still aim to minimize draw calls and state changes, however, it can be
+more difficult to batch together several objects into a single draw call. 3d
+meshes tend to comprise hundreds or thousands of triangles, and combining large
+meshes at runtime is prohibitively expensive. The costs of joining them quickly
+exceeds any benefits as the number of triangles grows per mesh. A much better
+alternative is to join meshes ahead of time (static meshes in relation to each
+other). This can either be done by artists, or programmatically within Godot.
+
+There is also a cost to batching together objects in 3d. Several objects
+rendered as one cannot be individually culled. An entire city that is off screen
+will still be rendered if it is joined to a single blade of grass that is on
+screen. So attempting to batch together 3d objects should take account of their
+location and effect on culling. Often though the benefits of joining static
+objects will outweigh other considerations, especially for large numbers of low
+poly objects.
```

Comment:
Despite this, the benefits of joining static objects will often outweigh other considerations